### PR TITLE
fix(DialogLayout): `footerSecondary` is only valid if `footer` is also specified

### PR DESCRIPTION
**Pull request by laraharrow**
- [x] 👾 Browsers tested
  - [x] Chrome
  - [x] Firefox
  - [x] Safari
  - [ ] IE11

### CODE REVIEW

File: packages/components/src/Dialog/Layout/DialogLayout.test.tsx [hunk 1]
```diff
@@ -58,4 +59,12 @@ describe('DialogLayout', () => {
...
     expect(screen.getByText('Cancel')).toBeInTheDocument()
   })
+  test('FooterSecondary without footer does not display', () => {
+    renderWithTheme(
+      <FooterSecondaryWithoutFooter {...FooterSecondaryWithoutFooter.args} />
+    )
+    expect(screen.getByText('Close')).toBeInTheDocument()
+    expect(screen.queryByText('Footer text')).not.toBeInTheDocument()
+    expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
+  })
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.test.tsx:69]: Include a test that "proves" that the type-union is preventing the developer from specifying `footerSecondary` without `footer`

See example at: https://github.com/looker-open-source/components/pull/2527/files#diff-70b3f612b78b582cb7e5380b31ecf51102774084bbfc3a482bca4dd532565d93R125-R132

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 2]
```diff
@@ -74,6 +74,15 @@ FooterSecondary.args = {
...
 }
 
+export const FooterSecondaryWithoutFooter = Template.bind({})
+FooterSecondaryWithoutFooter.args = {
+  ...Header.args,
+  footerSecondary: 'Cancel',
+}
+FooterSecondaryWithoutFooter.parameters = {
+  storyshots: { disable: true },
+}
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:84]: I think this should be causing a TS error?

File: packages/components/src/Dialog/Layout/DialogLayout.tsx [hunk 3]
```diff
@@ -35,31 +35,42 @@ import { DialogContent } from './DialogContent'
 import { DialogFooter } from './DialogFooter'
 import { DialogHeader, DialogHeaderProps } from './DialogHeader'
 
-export interface DialogLayoutProps extends ModalLayoutProps {
+type WithFooter = {
+  footer?: ModalLayoutProps['footer']
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.tsx:39]:
```suggestion
  footer: ModalLayoutProps['footer']
```

File: packages/components/src/Dialog/Layout/DialogLayout.tsx [hunk 4]
```diff
@@ -88,9 +107,7 @@ export const DialogLayout: FC<DialogLayoutProps> = ({
   headerDetail,
   isLoading,
 }) => {
-  const dialogFooter = (footer || footerSecondary) && (
-    <DialogFooter secondary={footerSecondary}>{footer}</DialogFooter>
-  )
+  const dialogFooter = constructDialogFooter(footer, footerSecondary)
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.tsx:110]: Given the simplicity of this logic I'd probably just inline it here:

```suggestion
  const dialogFooter = footer ? (
    <DialogFooter secondary={footerSecondary}>{footer}</DialogFooter>
  ) : null
```

File: packages/components/src/Dialog/Layout/DialogLayout.tsx (hunk 3 above)
Comment (lukelooker) [reply packages/components/src/Dialog/Layout/DialogLayout.tsx:39]: `footer` should also have a JSDoc explaining what it does: displays content on the right-side of the `DialogFooter`, uses `Space` to add whitespace between each element.

**Review by lukelooker — CHANGES_REQUESTED** [review]
Based on a read-through of the code it appears like the discriminated union isn't quite doing it's job yet. 

Added comments inline.

File: packages/components/src/Dialog/Layout/DialogLayout.tsx (hunk 3 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogLayout.tsx:39]: I don't think we can remove the `?` from footer's type.
That would indicate that every `DialogLayout` would need a `footer`

Comment (lukelooker) [reply packages/components/src/Dialog/Layout/DialogLayout.tsx:39]: With the optional undefined `?:` this type basically says that you can specify all the following without TS errors:

```
const testNone: FooterOptions = {}
const testFooter: FooterOptions = { footer: 'hello' }
const testFooterSecondary: FooterOptions = { footerSecondary: 'world' }
const testBoth: FooterOptions = { footer: 'hello', footerSecondary: 'world' }
```

But `testFooterSecondary` should cause a Typescript compile error.

The second `WithoutFooter` type allows for `footer` to be undefined entirely as long as `footerSecondary` is ALSO undefined.

**Review by lukelooker — APPROVED** [review]
Assuming tests pass this LGTM! :)

**Pull request merged**